flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinator.java

```diff
@@ -350,13 +340,15 @@ public CheckpointCoordinator(
                         this.minPauseBetweenCheckpoints,
                         this.pendingCheckpoints::size,
                         this.checkpointsCleaner::getNumberOfCheckpointsToClean);
+
         this.cachedTasksById =
-                new LinkedHashMap<ExecutionAttemptID, ExecutionVertex>(tasksToWaitFor.length) {
+                new LinkedHashMap<ExecutionAttemptID, ExecutionVertex>(
+                        attemptMappingProvider.getNumberOfTasks()) {
 
                     @Override
                     protected boolean removeEldestEntry(
                             Map.Entry<ExecutionAttemptID, ExecutionVertex> eldest) {
-                        return size() > CheckpointCoordinator.this.tasksToWaitFor.length;
+                        return size() > attemptMappingProvider.getNumberOfTasks();
```

Comment:
The interaction between cache in `CheckpointCoordinator` and `ExecutionAttemptMappingProvider` seems a bit fragile to me. And they both have kind of overlapping responsibilities.

What do you think about moving the cache istelf to the new `ExecutionAttemptMappingProvider`?

Reply:
I think it would be much more clear to move cache to `ExecutionAttemptMappingProvider`.  I then moved it to `ExecutionAttemptMappingProvider`.